function_queue.c

```diff
@@ -107,12 +117,19 @@ fq_pop(struct function_queue* q, struct function_queue_element* e, int block)
 		if(fq_is_empty(q, block) != 0) { /* underflow */
 			ret = ERANGE;
 		} else {
-			--q->size;
+			if( block )
+				ret = sem_wait(&q->size);

```

Comment:
If sem_wait() blocks here, the thread will block while holding the queue's mutex. This would prevent fq_push() from calling sem_post() (which also requires the mutex) so sem_wait() would not stop blocking. Condition variables might be a better solution.